src/test/fee_est/fee_est.cpp

```diff
@@ -0,0 +1,333 @@
+    int expectedBlocks = -1;
+    int actualBlocks = -1;
+};
+
+using TxMap = std::map<uint256, TxData>;
+
+const char* const GRAPH_HTML = R"(<!DOCTYPE html>
+<meta charset="utf-8">
+<title>Fee Graph</title>
+<script src="https://d3js.org/d3.v3.min.js"></script>
+<script src="https://cdnjs.cloudflare.com/ajax/libs/vega/3.0.0-beta.30/vega.js"></script>
+<script src="https://cdnjs.cloudflare.com/ajax/libs/vega-lite/2.0.0-beta.2/vega-lite.js"></script>
+<script src="https://cdnjs.cloudflare.com/ajax/libs/vega-embed/3.0.0-beta.14/vega-embed.js"></script>
+<style media="screen">
+.vega-actions a {
+    margin-right: 5px;
+}
+</style>
+<div id=vis></div>
+<script>
+var spec = {
+  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
+  "width": 800,
+  "height": 600,
+  "layer": [
+    {
+        "data": {
+            "values": %s
+        },
+        "mark": "tick",
+        "encoding": {
+            "x": {"field": "blocks", "type": "quantitative"},
+            "y": {"field": "feeRate", "type": "quantitative"},
+            "color": {"field": "height", "type": "quantitative"}
+        }
+    },
+    {
+      "data": {
+        "values": %s
+      },
+      "encoding": {
+        "x": {"field": "blocks", "type": "quantitative"},
+        "y": {"field": "feeRate", "type": "quantitative"},
+        "color": {"field": "mode", "type": "nominal"}
+      },
+      "mark": "line",
+    }
+  ]
+}
+vega.embed("#vis", spec);
+</script>
+)";
+
+// Maximum number of randomly sampled transactions to show on graph.
+const int SAMPLE_TXS = 50000;
+
+// Drop 1/20th of sampled transactions with highest feerates if they are
+// outliers that would change the scale of the graph.
+const int DROP_FRAC = 20;
+
+void WriteGraph(const std::string& filename,
+    const TxMap& txMap,
+    const std::vector<unsigned int>& targets,
+    CBlockPolicyEstimator& estimator)
+{
+    UniValue fees(UniValue::VARR);
+
+    int64_t maxFeeRate = 0;
+    const unsigned int maxTarget = estimator.getMaxTarget();
+    for (bool conservative : {false, true}) {
+        for (unsigned int target : targets) {
+            if (target > maxTarget) continue;
+            UniValue fee(UniValue::VOBJ);
+            fee.pushKV("mode", conservative ? "conservative" : "default");
+            fee.pushKV("blocks", int(target));
+            int64_t feeRate = estimator.estimateSmartFee(target, nullptr /* confTarget */, conservative).GetFeePerK();
+            fee.pushKV("feeRate", feeRate);
+            fees.push_back(fee);
+            maxFeeRate = std::max(maxFeeRate, feeRate);
+        }
+    }
+
+    std::vector<const TxMap::value_type*> sample;
+    sample.reserve(SAMPLE_TXS);
+    int i = 0;
+    std::minstd_rand randint;
+    for (const auto& entry : txMap) {
+        if (entry.second.actualBlocks > int(maxTarget)) {
+            continue;
+        } else if (sample.size() < SAMPLE_TXS) {
+            sample.emplace_back(&entry);
+        } else {
+            int j = randint() % i;
```

Comment:
An implicit signedness changing conversion takes place here. Avoid or make it explicit :-)